### ISSUE TITLE
RFC: Add option to disable ohmyzsh aliases

### ISSUE THREAD

**Issue by carlosala** [ohmyzsh/ohmyzsh#10644]
Hi everyone!

It's a common sentiment to feel that omz adds too many aliases in different areas, such as `git` ones (discussed in #5721), default omz ones (#9414) or others. There's a pull request actually (#10510) that fix this issue, but it uses environment variables (not recommended actually, as it pollutes the global variable namespace) and is not granular, as you only can disable all default omz aliases and all plugin aliases.

I would like to ask you what do you think about this solution, as I think that is cleaner:

Plugin aliases would be disabled by setting the following snippet in your `.zshrc`:

```zsh
zstyle ':omz:plugins:<nameofplugin>' alias off
```

For instance:

```zsh
zstyle ':omz:plugins:yarn' alias off
```

They can also be disabled all at once, using wildcards:

```zsh
zstyle ':omz:plugins:*' alias off
```

And enable only a few plugins:

```zsh
zstyle ':omz:plugins:*' alias off
zstyle ':omz:plugins:yarn' alias on
zstyle ':omz:plugins:git' alias on
```

For default ohmyzsh aliases we could use a similar way of doing it, and disabling the whole set of aliases would be something similar to:

```zsh
zstyle ':omz:*:*' alias off
```

What do you think about this idea? Let's discuss it here!

**Comment by remcohaszing** [comment]
This sounds great! Maybe you could take it a step further and use `zstyle` to replace the `plugins` array and other variables entirely?

```diff
- ZSH_THEME="risto"
- DISABLE_AUTO_UPDATE='true'
- plugins=(
-   colored-man-pages
-   git
- )
+ zstyle ':omz:theme' risto
+ zstyle ':omz:auto-updated' off
+ zstyle ':omz:plugins:colored-man-pages' enabled on
+ zstyle ':omz:plugins:git' enabled on

  source "$ZSH/oh-my-zsh.sh"
```

I’m not familiar with `zstyle` so I don’t know what the possibilities are.

**Comment by njfamirm** [comment]
Can't we get aliases as a list from plugins first, And by iterating, add them?

The format of these lists is the same as OMZ_ALIAS_pluginname
if user want to disable alias of plugin, easily we `unset` list, before iterating

.zshrc
```zsh
disables_alias=(OMZ_ALIAS_git OMZ_ALIAS_golang)
```
plugin.zsh
```zsh
declare -A omzaliases
omzaliases=([home]=cd)
```
oh-my-zsh.sh
```zsh
# first unset disables alias
for d in $disables_alias; do
  unset $d
done

# add aliases
for key in "${(k)omzaliases}"; do
  alias $key=$omzaliases[$key]
done
```
*This script is very basic and should be expanded*

**Comment by Drumstickx** [comment]
I like the granularity of the approach. I considered an extension for per-alias granularity but am not sure whether the additional complexity justifies the potential use cases. However, I share my thoughts for the sake of discussion.

Potential extension for per-alias granularity:

```
zstyle ':omz:plugins:<nameofplugin>:<nameofalias>' alias <off/on>
```

Use Cases:

#### 1. Enabling a subset of aliases of a plugin

- Issue: Some plugins define a lot of aliases that clutter the command namespace
- Solutions:
  - Unsetting all other aliases manually via `unset` after loading `ohmyzsh`
    - Issue: Very cumbersome, newly added aliases have to be added manually to the list of excluded aliases
  - (with per-plugin granularity) Disabling all aliases for the plugin and manually declaring a subset of aliases
    - Issue: Manually declared aliases are not updated if their counterparts in the plugin code are
  - (with per-alias granularity) Disabling the alias for all plugins where it should not be set

#### 2. Disabling a subset of aliases of a plugin

- Issue: The aliases of certain plugins may conflict with commands or aliases of other programs
- Solutions:
  - Unsetting the aliases manually via `unset` after loading `ohmyzsh`
    - Acceptable solution (in my opinion)
  - (with per-plugin granularity) Disabling all aliases for a plugin with conflicting aliases
    - Issue: Other aliases cannot be used unless they are manually declared
  - (with per-alias granularity) Each unwanted alias can be disabled without affecting other aliases of the plugin

#### 3. Duplicate aliases (e.g. `acs` in the `aliases` and `debian` plugins)

- Issue: When multiple plugins specify the same alias, the plugin instantiated last "wins"  and its definition of the alias is used while another definition may be desired
- Solutions:
  - Changing the order of conflicting plugins
    - Issue: I prefer the list of plugins to be sorted alphabetically
  - Manually overwriting the alias via re-declaration after loading `ohmyzsh`
    - Issue: Manually declared aliases are not updated if their counterparts in the plugin code are
  - (with per-plugin granularity) Disabling all plugins that overwrite the desired alias
    - Issue: All aliases of plugins with conflicting aliases are disabled
  - (with per-alias granularity) Disabling the alias for all plugins where it should not be set

**Comment by b-hayes** [comment]
I just want all of them gone I hate them. Coming from bash a lot of zsh completions are wack already and the aliases just add way too much noise.
Is there a way to turn them off with a workaround while we wait?

**Comment by rwmitchell** [comment]
What plugins do you use?

There are only a few aliases defined in the base config, most are are in the add on plugins, like ‘git’

**Comment by b-hayes** [comment]
this is in my zshrc `plugins=(git docker docker-compose)`
I commented the plugins line out to see the difference and still get a bunch of extra junk eg git check[TAB]:
<img width="596" alt="image" src="https://user-images.githubusercontent.com/30540495/176592673-ca68b4d2-d3bc-470a-bf48-a87b8021ff80.png">

**Comment by carlosala** [comment]
Those are commands from git, omz has nothing to do with that!

**Comment by b-hayes** [comment]
Weird. Ive never seen git have those commands before coming to macos and zsh.

**Comment by rwmitchell** [comment]
It is part of zsh completions, has nothing to do with aliases. Type ```alias``` to see defined aliases.

**Comment by remcohaszing** [comment]
omz contains two plugins for git completion:

1. `git` contains full completion for all git commands
2. `gitfast` only contains completions for some commonly used git commands

**Comment by b-hayes** [comment]
Thanks @remcohaszing !

**Comment by azinsharaf** [comment]
Checking on this request, is it possible to disable aliases now?

**Comment by rwmitchell** [comment]
Which aliases do you want disabled?

**Comment by azinsharaf** [comment]
I would say all of them. And enable the ones i like and add my own aliases.

**Comment by remcohaszing** [comment]
You can append the plugins to `fpath`.

```zsh
# You probably declared this variable somewhere
ZSH="$XDG_DATA_HOME/oh-my-zsh"

# …

# I placed this before sourcing oh-my-zsh.sh, but I don’t think it really matters
fpath+="$ZSH/plugins/adb"
fpath+="$ZSH/plugins/docker"
fpath+="$ZSH/plugins/pip"
fpath+="$ZSH/plugins/yarn"

# …

source "$ZSH/oh-my-zsh.sh"
```

**Comment by mcornella** [comment]
This was implemented in #11550, see https://github.com/ohmyzsh/ohmyzsh#skip-aliases for how to use it. Thanks everyone for your comments!
